Add restore handler for password input

grisu48 opened the pull request:
This commit restores the terminal state in case the program is interrupted while being in password read mode. This ensures the terminal remains usable, also if the password input is being cancelled

This PR provides a solution for https://github.com/containers/podman/issues/17372

Signed-off-by: Felix Niederwanger <felix.niederwanger@suse.com>

grisu48 commented:
I'm not sure why the DCO fails. The commit and PR message have both the required `Signed-off-by` footer appended :thinking:

Luap99 commented:
The DCO is failing because you did not set your actual name for the git commit (`git config user.name`). But I don't think this is required for us, we can set the check manually to pass.

Luap99 commented on pkg/auth/auth.go:
```diff
@@ -244,7 +245,25 @@ func getUserAndPass(opts *LoginOptions, password, userFromAuthFile string) (user
+		if err != nil {
+			oldState = nil
+		}
+		sigch := make(chan os.Signal, 1)
```
the channel must be closed in a defer call otherwise the goroutine will leak forever.

grisu48 replied:
Good catch. I reset the signal interrupt handler but forgot the channel.

Commit amended.

grisu48 commented:
> The DCO is failing because you did not set your actual name for the git commit (`git config user.name`). But I don't think this is required for us, we can set the check manually to pass.

Ahhhhhhh, thanks. Fixed and force-pushed.

Luap99 commented on pkg/auth/auth.go:
```diff
@@ -244,7 +245,26 @@ func getUserAndPass(opts *LoginOptions, password, userFromAuthFile string) (user
+				if oldState != nil {
+					terminal.Restore(stdin, oldState)
+				}
+				os.Exit(1)
```
I don't think we want to call exit so far down in the stack. At least in podman we have handlers that depend on proper function, see `libpod/shutdown`.

Also I have no idea where this code is used besides podman, I think it might be better to return an error so that the caller can decide what to do with it.

grisu48 replied:
I'm not sure how to handle this - In the podman repository I couldn't find any signal handlers for `os.Interrupt` except for [machine.go](https://github.com/containers/podman/blob/main/pkg/machine/qemu/machine.go#L491), which also uses `os.Exit(1)`.

I can certainly modify it such that `getUserAndPass` returns a `fmt.Errorf("interrupted")` but I'm not sure if this is the desired behaviour. What do you think?

vrothberg replied:
I concur that we should return an error here.

Luap99 reviewed (commented):
@vrothberg @mtrmac PTAL

vrothberg reviewed (commented):
Thanks for contributing!

It's probably worth moving the logic around `terminal.ReadPassword` into a separate function and update all callers (incl. pkg/ssh).

grisu48 commented:
I've updated the PR where I refactored a `ReadPassword` function to `utils/utils.go`. There are some places that would need to be updated later on as well, as @vrothberg stated.  
They are located in other repositories and require this PR first, but it can be easily done afterwards.

mtrmac commented on pkg/util/util.go:
```diff
@@ -22,3 +29,31 @@ func StringMatchRegexSlice(s string, re []string) bool {
+			if oldState != nil {
+				terminal.Restore(stdin, oldState)
+			}
+			err = fmt.Errorf("interrupted")
```
AFAICS this doesn’t work at all: The original tiny example has an `os.Exit`, which we don’t want.

But doing _nothing_ here would not interrupt `terminate.ReadPassword` at all; in my experiments the program just starts echoing Ctrl-C (because of the `.Restore`) but `ReadPassword` never exits. And I can’t see any plausible way to terminate `ReadPassword`, except maybe for closing the `stdin` FD, which would just result in mayhem (because the FD would be reassigned by any future file open)

grisu48 replied:
This is very true. Please see my comment below for advice on how to proceed.

mtrmac commented:
> In the extreme, we could fork / spawn a subprocess. Ugh… 

That wouldn’t, alone be sufficient either, because Ctrl-C causes `SIGINT` to be sent to the whole foreground process group, including bot the child and the parent, so Podman would still need to arrange to ignore `SIGINT` in some reversible way. And Podman creating a separate process group is another can of worms… that would interfere with Ctrl-Z / `fg` / `bg`.

I wouldn’t _entirely_ rule out the possibility of making all of that work, but the complexity (and dependency one of the most obscure aspects of POSIX — with Windows and other systems a whole another can of worms) makes it _very_ attractive to me to suggest to just give up on this. Convincing users to use a more modern / featured shell that resets the terminal state before printing a prompt, or providing such patches to major shells, is starting to look easier than Podman handling this…

---

One, sort of, alternative is to try to make the `os.Exit` approach more viable, by explicitly introducing some kind of ”CLI parsing phase” to Podman, where Podman is not yet allowed to really do anything substantial (but what about user namespaces and `reexec??), and providing only a `ReadPassprhaseInCLIParsingPhase` that makes a lot of assumptions about the environment (it’s safe to call `os.Exit`; this is the only package so far to invoke `os/signal`; and the like).

But, again, ugh.

grisu48 commented on pkg/util/util.go:
```diff
@@ -22,3 +29,31 @@ func StringMatchRegexSlice(s string, re []string) bool {
 }
+
+// ReadPassword reads a password from the terminal and restores the terminal state on interruption
+func ReadPassword() (string, error) {
```
Done.

grisu48 commented on pkg/util/util.go:
```diff
@@ -22,3 +29,31 @@ func StringMatchRegexSlice(s string, re []string) bool {
+	if err != nil {
+		oldState = nil
+	}
+	sigch := make(chan os.Signal, 1)
```
Done.

grisu48 commented on pkg/util/util.go:
```diff
@@ -1,6 +1,13 @@
+	"os/signal"
+	"regexp"
+
+	terminal "golang.org/x/term"
```
Done.

grisu48 commented:
Thanks for the review and the input! I see two paths on how this could be addresses, both have their own caveats:

1. Don't stop the INTERRUPTED signal, just restore the terminal

Please see suggestion 1 in the PR. I'm catching SIGINT, restore the terminal and then re-emit the signal. This is IMHO the least invasive change, as it should not alter the existing behaviour much: It would only restore the terminal as intermittent step.

2. Handle SIGINT gracefully, return an error from `ReadPassword` but leave the terminal reader active, and thus stdin blocked

This is ugly, but has the advantage, that `podman` can run it's clean-up routines, as suggested by @Luap99 and @vrothberg. Successive access to stdin would result in unintended behaviour though, so this might be better from the perspective of `ReadPassword` but introduces a tripping hazard.

Please see the current commit, where I added a suggestion for both methods.

***

How should we continue? I have a preference for solution 1, but would leave it to someone with far better insight to suggest on how to proceed.

Thank you!

grisu48 commented on pkg/util/util.go:
```diff
@@ -22,3 +29,31 @@ func StringMatchRegexSlice(s string, re []string) bool {
+			if oldState != nil {
+				terminal.Restore(stdin, oldState)
+			}
+			err = fmt.Errorf("interrupted")
```
Done, I've introduced another channel for this.

grisu48 commented on pkg/util/util.go:
```diff
@@ -22,3 +29,31 @@ func StringMatchRegexSlice(s string, re []string) bool {
+		}
+	}()
+	password, err = terminal.ReadPassword(stdin)
+	signal.Reset(os.Interrupt)
```
Done. Changed to stop and added `defer`

grisu48 commented on pkg/util/util.go:
```diff
@@ -22,3 +29,31 @@ func StringMatchRegexSlice(s string, re []string) bool {
 }
+
+// ReadPassword reads a password from the terminal and restores the terminal state on interruption
+func ReadPassword() (string, error) {
```
Done, changed to `([]byte, error)`

mtrmac commented on pkg/util/util.go:
```diff
@@ -22,3 +30,64 @@ func StringMatchRegexSlice(s string, re []string) bool {
+					term.Restore(fd, oldState)
+				}
+				signal.Stop(interruptChannel)
+				syscall.Kill(syscall.Getpid(), syscall.SIGINT)
```
Suggestion 1:

This works in a top-level application; it’s fairly suspect in library code. E.g. if there is anything else registered via `signal.Notify`, it will now see the signal twice.

Also, even with this approach, I think we _need_ to interrupt `term.ReadPassword`, because there could be some other party with a signal handler; in that case even this `syscall.Kill` might never terminate the `term.ReadPassword`.

The good thing about this approach is that the exit code of the process actually reflects the signal. (BTW that doesn’t actually happen directly here, there is _still_ a Go signal handler installed, but it does ~the same thing, uninstalling the handler and re-sending the signal.)

grisu48 replied:
My main concern is that there is no way of determining if there are other signal handlers for SIGINT present that attempt to gracefully terminate the application. This signal handler re-emits SIGINT and kills them. I could not find anything in `os/signal` that allows us to check, if there are other handlers present.

This makes this solution undesirable, as it would kill possible other signal handlers, apart from your points.

mtrmac commented on pkg/util/util.go:
```diff
@@ -22,3 +30,64 @@ func StringMatchRegexSlice(s string, re []string) bool {
+		defer close(inputChannel)
+		go func() {
+			var buf ByteBuffer
+			buf.buf, buf.err = term.ReadPassword(fd)
+			inputChannel <- buf
```
Suggestion 2:

With this, `term.ReadPassword` continues running, consuming resources (meh), but potentially interfering: it continues to consume data from stdin, and it can eventually return and change the terminal state at any future time.

I can’t see how we can terminate the current implementation: `fd` must be a terminal (so that ioctls work); and I can’t see that a pending `read(3)` would be interrupted by a `close(3)`.

---

A downside of turning the signal into an error is that the exit status of the process will not be traditionally correct. (In principle, we could turn it into a recognizable error, and turn that error into a signal at the very top of Podman.)

grisu48 replied:
Yes. The only way of solving this would be to change `term.ReadPassword` to handle interrupts itself gracefully.

mtrmac reviewed (commented):
I really think we can’t reasonably solve this locally in c/common.

90% of the problem is that `golang.org/x/term.ReadPassword` _needs_ to be interruptible (maybe using a `context.Context`). That would make its existing implementation automatically take care of restoring the TTY configuration.

Then we would need to “only” care about turning a signal into a cancellation + error, and possibly turning the error back into a signal. That’s still not _trivial_ — but having the cancellation feature would allow us to turn the signal into a cancellation+error using a _very_ non-invasive` signal.Notify` which doesn’t interfere with anything else in the process, and to turn the error into a signal at the very top of the Podman call stack, where there is nothing else to interfere with.

So, would anyone care to propose this to `…/x/term` maintainers?

mtrmac commented:
> (x/term does have some code that attempts to handle Ctrl-C, 

My mistake: `term.ReadPassword` and `term.Terminal.ReadPassword` are completely independent implementations, and the quoted note refers to the latter, which we don’t use. (Using it would not help us anyway...)

grisu48 commented on pkg/util/util.go:
```diff
@@ -22,3 +30,64 @@ func StringMatchRegexSlice(s string, re []string) bool {
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, os.Interrupt)
+		defer signal.Stop(interruptChannel)
+		defer close(interruptChannel)
```
Done

grisu48 commented:
The issue is filed in https://github.com/golang/go/issues/31180 and so far there has not been much activity or any visible attempts to solve this.

If we would get an interruptible `term.ReadPassword` implementation, then suggestion 2 could be changed to a clean library function.

Looks like we're hitting a wall here.

mtrmac commented:
My suggestion for what to try would be to file a PR directly at https://github.com/golang/term , to add something like `ReadPasswordWithContext`. But I have no idea whether it would be likely to be accepted, and it’s also not work I can prioritize doing myself.

grisu48 commented:
Thanks. I'll have a look and see what we can do. I leave this PR open and update it accordingly, so we keep track of the progress.

grisu48 commented:
I'm working on a prototype for `ReadPasswordWithContext` but am busy with other tasks as well, so this might take some time.

mtrmac commented:
> Also: This requires a Unix/Windows implementation on a rather low level, so this is a nontrivial task.

From a _very_ quick look it seems that it could be possible to wrap the `Reader` passed to `readPasswordLine` with one that checks whether a context was canceled, but I didn’t actually try implementing that so I might be very off-base.

grisu48 commented:
> > Also: This requires a Unix/Windows implementation on a rather low level, so this is a nontrivial task.
> 
> From a _very_ quick look it seems that it could be possible to wrap the `Reader` passed to `readPasswordLine` with one that checks whether a context was canceled, but I didn’t actually try implementing that so I might be very off-base.

The problem with this approach is that a ongoing `Read` from the reader would block until the next character, even if the context is cancelled. I try to avoid this issue by implementing a nonblocking polling approach for this, which should in principle not have this limitation. Implementing this for all supported platforms, including Windows, is a however not as easy as I was hoping for AFAICS :slightly_smiling_face:

grisu48 commented:
No reply for two weeks on https://github.com/golang/term/pull/12. I also believe this is not going to fly because there two main issues that can not be resolved elegantly within the library function:

1. Readers can't be interrupted. I solve this by non-blocking IO, but that's ugly and it does not work on Windows (hmpf!)
2. Non-blocking IO requires polling, which or hogs the CPU or adds a (hopefully not noticeable) latency

I believe we're hitting a dead end there as well. Has anyone an alternative idea how we would approach this issue?  
I would still argue, that having a signal handler which restores the terminal and then calls `os.Exit` might be the best of the bad options, given that this is a) AFAICS the current behaviour and b) missing alternatives.

What do you think?

mtrmac commented:
I think that leaves the `ReadPassphraseInCLIParsingPhase` approach… Aesthetically I find that quite unattractive, but I’ll leave it to Podman maintainers to decide whether it is is worth it.

grisu48 commented:
Reviving this PR.

Please review my current suggestion. After multiple attempts we concluded that there is no perfect solution, as `ReadPassword` from `x/term` leaves the terminal in non-echo mode (See https://github.com/golang/go/issues/31180).

My suggestion based on https://github.com/containers/common/pull/1312#issuecomment-1483308019 is to implement a special `ReadPassphraseInCLIParsingPhase` function, which is allowed to terminate `podman` after restoring the terminal state. Since this function is only used in the authentication, where the current behaviour of `podman` is to terminate due to SIGINT, this would not alter the current program flow.

Other attempts to do busy waiting on a non-blocking terminal might work as well but have their own caveats (increased system load and latency due to busy polling). Also they don't work on Windows.

Asking @vrothberg @Luap99 on their opinion on this topic or if they have a better idea how to approach this issue.

vrothberg commented on pkg/util/util_supported.go:
```diff
@@ -89,3 +91,28 @@ func GetRuntimeDir() (string, error) {
+				terminal.Restore(fd, oldState)
+			}
+			signal.Reset(os.Interrupt)
+			os.Exit(1)
```
Why not just return an error and give Podman a chance to properly exit?  There is usually some cleanup do be done before Podman exists.

grisu48 replied:
Because there is no way to return from `terminal.ReadPassword`, so this would mean the terminal remains blocked.

rhatdan replied:
So if you returned the error, users would need to hit <return> or if you exit here users have to hit <return> correct?

grisu48 replied:
Not sure that I understand you - If we return a error, the stdin would still be blocked by `terminal.ReadPassword` until users hit return.

We even [tried to](https://github.com/containers/common/pull/1312#discussion_r1099038041) close the fd but not even that unblocks the `read` call from `terminal.ReadPassword` . So returning an error here on interrupt is possible, but it would mean that the terminal remains in a weird state, which is likely not usable on subsequent calls. :slightly_frowning_face:

rhatdan replied:
Have you experimented with returning, and seen what happens from the user point of view?

grisu48 replied:
Yes and this is interesting because it revealed a race condition. There is somewhere else a signal handler for the interrupt, which terminates `podman` with "exit status 1". We would need to disable this signal handler for this function to work properly, which is ugly, but it would work. From the user's perspective there is no noticeable change, except that the return code would be 125 instead of 1.

Should we switch to that approach? It's ugly because we would need to reset the SIGINT signal handler and in the error case the terminal is blocked until the program terminates, but the alternative is to use `os.Exit(1)` within the function, which is nasty for other reasons.

I have no preference and am happy to continue with either way :slightly_smiling_face:

mheon replied:
The SIGINT handler in question should be https://github.com/containers/podman/blob/main/libpod/shutdown/handler.go but that doesn't help us much - that's nestled deep into Libpod and as such we can't touch it from here in c/common.

Generally speaking we should avoid `os.Exit()` whenever possible (for the exact reasons we wrote the shutdown handler in Libpod - there are situations where we need to keep Podman alive long enough to finish essential cleanup before obeying user-initiated shutdown), so I'm in favor of returning an error, but I have no easy way of disabling the interrupt handler from within c/common to give us consistent error codes.

grisu48 replied:
I see that almost everyone in this threads favours an approach of returning an error. This doesn't remove the race condition with [libpod/shutdown/handler.go](https://github.com/containers/podman/blob/0b612a9a2d812308e892dd909c53b4b2e0cfde3d/libpod/shutdown/handler.go#L45C57-L45C57), but it works in most of the cases I tested, so it would improve the situation already substantially.

I will update the PR accordingly.

Luap99 replied:
So I think the answer here is to take a step back, I understand what this PR is trying to do and that looks reasonable. 

If we look at the code path exactly `podman login` bypasses pretty much all of libpod and calls directly in to pkg/auth here. So shouldn't be the answer disable the libpod setup for podman login/logout in which case there cannot be a race between the signal handlers.

rhatdan commented:
/approve
LGTM
@vrothberg @Luap99 @mtrmac PTAL

Luap99 commented on pkg/util/util_supported.go:
```diff
@@ -89,3 +91,45 @@ func GetRuntimeDir() (string, error) {
+			if oldState != nil {
+				_ = terminal.Restore(fd, oldState)
+			}
+			errorChannel <- Buffer{Buffer: make([]byte, 0), Error: fmt.Errorf("interrupted")}
```
if we return an error it should return a typed error that the client can catch with errors.Is()

grisu48 replied:
Good point! I've added a [`ErrInterrupt` in utils.go](https://github.com/containers/common/pull/1312/files#diff-0fa449830b5dc2b274b7c95a58b5071edc23f1e44af84a99c1fa9e600767829fR24)

rhatdan commented:
@giuseppe @Luap99 @vrothberg @mtrmac PTAL

grisu48 commented:
Rebased and applied `gofumpt`.

grisu48 commented:
@rhatdan ping?

rhatdan commented:
LGTM
@giuseppe @Luap99 @vrothberg @mtrmac PTAL

vrothberg reviewed (approved):
/lgtm
